sled-agent/src/instance_manager.rs

```diff
@@ -199,6 +222,316 @@ impl InstanceManager {
+        instance_id: Uuid,
+        old_runtime: &InstanceRuntimeState,
+        migration_ids: &Option<InstanceMigrationSourceParams>,
+    ) -> Result<SledInstanceState, Error> {
+        let (tx, rx) = oneshot::channel();
+        self.inner
+            .tx
+            .send(InstanceManagerRequest::PutMigrationIds {
+                instance_id,
+                old_runtime: old_runtime.clone(),
+                migration_ids: *migration_ids,
+                tx,
+            })
+            .await
+            .map_err(|_| Error::FailedSendChannelClosed)?;
+        rx.await?
+    }
+
+    pub async fn instance_issue_disk_snapshot_request(
+        &self,
+        instance_id: Uuid,
+        disk_id: Uuid,
+        snapshot_id: Uuid,
+    ) -> Result<(), Error> {
+        let (tx, rx) = oneshot::channel();
+        self.inner
+            .tx
+            .send(InstanceManagerRequest::InstanceIssueDiskSnapshot {
+                instance_id,
+                disk_id,
+                snapshot_id,
+                tx,
+            })
+            .await
+            .map_err(|_| Error::FailedSendChannelClosed)?;
+        rx.await?
+    }
+
+    /// Create a zone bundle from a named instance zone, if it exists.
+    pub async fn create_zone_bundle(
+        &self,
+        name: &str,
+    ) -> Result<ZoneBundleMetadata, BundleError> {
+        let (tx, rx) = oneshot::channel();
+        self.inner
+            .tx
+            .send(InstanceManagerRequest::CreateZoneBundle {
+                name: name.to_string(),
+                tx,
+            })
+            .await
+            .map_err(|err| BundleError::FailedSend(anyhow!(err)))?;
+        rx.await.map_err(|err| BundleError::DroppedRequest(anyhow!(err)))?
+    }
+
+    pub async fn add_external_ip(
+        &self,
+        instance_id: Uuid,
+        ip: &InstanceExternalIpBody,
+    ) -> Result<(), Error> {
+        let (tx, rx) = oneshot::channel();
+        self.inner
+            .tx
+            .send(InstanceManagerRequest::InstanceAddExternalIp {
+                instance_id,
+                ip: *ip,
+                tx,
+            })
+            .await
+            .map_err(|_| Error::FailedSendChannelClosed)?;
+        rx.await?
+    }
+
+    pub async fn delete_external_ip(
+        &self,
+        instance_id: Uuid,
+        ip: &InstanceExternalIpBody,
+    ) -> Result<(), Error> {
+        let (tx, rx) = oneshot::channel();
+        self.inner
+            .tx
+            .send(InstanceManagerRequest::InstanceDeleteExternalIp {
+                instance_id,
+                ip: *ip,
+                tx,
+            })
+            .await
+            .map_err(|_| Error::FailedSendChannelClosed)?;
+        rx.await?
+    }
+}
+
+// Most requests that can be sent to the "InstanceManagerRunner" task.
+//
+// These messages are sent by "InstanceManager"'s interface, and processed by
+// the runner task.
+//
+// By convention, responses are sent on the "tx" oneshot.
```

Comment:
Yup, definitely inspired by your code using mpsc + oneshot. I'm using `tx` because of how universal the replies are, so I don't *think* it's ambiguous right now.